### ISSUE TITLE
Building database takes very long

### ISSUE THREAD

**Issue by mdehollander** [pirovc/ganon#50]
This tool got my attention, because I hope it will make maintaining reference databases a lot easier.

I installed Ganon from this github repository and started to build a database (fungal genomes, 60 GB). According to the information in the abstract of the pre-print, this should take 1 hour or so. But now even after a day it is still busy with the taxsbp script get_len_taxid.sh. Things are still being written to `acc_len_taxid.txt`. I see that the scripts queries the ncbi via eutils. Could it be that this step is rate limited? Any suggestions to make this run faster? I think tools like kraken download the taxonomy and than parse it locally.

This is how I ran Ganon:
```
$ ./ganon build --ganon-path build/ --taxsbp-path ../taxsbp/ -d ganon-genbank-fungi -t 8 -i /data/db/genomes/genbank/fungi/*.gz --verbose                                  
Extracting accessions... Done. Elapsed time: 1212.7593030929565 seconds.
```

**Comment by pirovc** [comment]
Hi @mdehollander 

Nice, ganon is indeed being delevoped with database maintenance in mind.

Such step sometimes can take a while since NCBI is limiting users to have only 3 request each second. For such reason I recently updated the script, now with request in batches. Make sure you are using this version: https://github.com/pirovc/taxsbp/blob/v0.1/scripts/get_len_taxid.sh

Let me know if that solves your problem, otherwise we can figure another way out.

PS: ganon is in constant changes due to final developments to a first release, so it may happen that small things won't work or change in the course of next weeks. Soon I will release the first official version and  upload the manual, so things will be more clear.

**Comment by mdehollander** [comment]
I am on on the 0.1 branch and I see `batch=200` in the script. So this this did not solve the issue. I have around 1700 genomes that I would like to index.

I am aware that Ganon is still under development. I love to try out new software and the preprint looks promising. And yes, a manual would be great :)

**Comment by pirovc** [comment]
Are all the genomes with the standand >accession.version header? are your sequences from a manually curated database or just downloaded from directly genbank?

1700 should take just a few minutes. It can be a bug in the script since I just fixed it in the last weeks or NCBI is holding you back. Hard to tell. If I have access to the accessions I could try it out.

If you are using a standard database, I can recomend you to use another script I developed to download sequences: https://github.com/pirovc/genome_updater

`./genome_updater.sh -d genbank -g fungi -f "genomic.fna.gz,assembly_report.txt" -m -u -r -p -t 8`

with that you can easily collect the data needed for ganon. Actually this script will be incorporated to ganon later on, since it can also update downloaded files

**Comment by mdehollander** [comment]
The files are with standard header. I downloaded directly from NCBI with aspera.
As a check I used your script. I added that I only want representative genomes:
```
$ ./genome_updater.sh -d genbank -g fungi -c "representative genome" -f "genomic.fna.gz,assembly_report.txt"
 -m -u -r -p -t 8                                             
----------------------------------------
      genome_updater version: 0.10
----------------------------------------
-- NEW --
Downloading assembly summary [2019-02-20_15-51-10_assembly_summary.txt]...
 - 2492 out of 4329 entries removed [RefSeq category: representative genome, Assembly level: all, Version status: latest]
 - 1837 entries available
 - Downloading 3674 files with 8 threads...
/bin/bash: line 4: printf: .0200: invalid number
/bin/bash: line 4: printf: .1900: invalid number
/bin/bash: line 4: printf: .1000: invalid number
```
As you see, there is a print error, but I think that is not crucial, only for summary information or not? Do you need a new issue on the genome_updater page for it?

When I now run Ganon, I see the same behaviour:
```
./ganon build --ganon-path build/ --taxsbp-path ../taxsbp/ -d ganon-genbank-fungi -t 8 -i ~/install/genome_updater/db/files/*.gz
Extracting accessions... Done. Elapsed time: 1180.3812501430511 seconds.
```
```
$ wc -l accessions.txt 
6194973 accessions.txt
$ wc -l acc_len_taxid.txt 
30200 acc_len_taxid.txt
```
So getting the accession and the length goes very slow. I see some `curl` processes going to eutils, but no fast proges. My internet connection is fine, could NCBI rate limit the ip I have? 
How long does this take for you? Any suggestions what I can do to improve this?

**Comment by pirovc** [comment]
Actually now that you have data from genome_updater you could simply do :

`awk -F "\t" 'BEGIN {FS="\t";OFS="\t"}{if($4!="na"){ print $4,$5,$6,$2 }}' download_folder/TIMESTAMP*_updated_sequence_accession.txt > acc_len_taxid_assembly.txt` 

changing the `download_folder/TIMESTAMP` values to the ones you have.

When running ganon, add the following parameter:

`--len-taxid-file acc_len_taxid_assembly.txt`

that will skip the slow pre-processing steps.

It would be nice if you report on genome_updater as well.
I guess NCBI is holding you back, but now that I can reproduce it we'll see.

**Comment by mdehollander** [comment]
Thanks! This speed things up. Great. I had to remove a 'na' number of `acc_len_taxid_assembly.txt` because for one genome the length could not be retrieved or something like that.

I think NCBI did not like that the build tool was querying a lot, since we got a ban today:
![image](https://user-images.githubusercontent.com/262872/53182998-d8636700-35fa-11e9-95bf-4ea2b6dc1eef.png)
Although the script is using the e-utils, it seems to violate the NCBI fair-use policy. Using the awk script seems a better way to do this.
